In [19]:
import pandas as pd

We will first read the csv file titled RAW_recipes.

In [20]:
RAW_recipes=pd.read_csv('RAW_recipes.csv')

In [21]:
type(RAW_recipes)

pandas.core.frame.DataFrame

In [22]:
RAW_recipes.head()

name      id  minutes  \
0  arriba   baked winter squash mexican style  137739       55   
1            a bit different  breakfast pizza   31490       30   
2                   all in the kitchen  chili  112140      130   
3                          alouette  potatoes   59389       45   
4          amish  tomato ketchup  for canning   44061      190   

   contributor_id   submitted  \
0           47892  2005-09-16   
1           26278  2002-06-17   
2          196586  2005-02-25   
3           68585  2003-04-14   
4           41706  2002-10-25   

                                                tags  \
0  ['60-minutes-or-less', 'time-to-make', 'course...   
1  ['30-minutes-or-less', 'time-to-make', 'course...   
2  ['time-to-make', 'course', 'preparation', 'mai...   
3  ['60-minutes-or-less', 'time-to-make', 'course...   
4  ['weeknight', 'time-to-make', 'course', 'main-...   

                                    nutrition  n_steps  \
0       [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]       11   
1   [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]        9   
2  [269.8, 22.0, 32.0, 48.0, 39.0, 27.0, 5.0]        6   
3   [368.1, 17.0, 10.0, 2.0, 14.0, 8.0, 20.0]       11   
4   [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]        5   

                                               steps  \
0  ['make a choice and proceed with recipe', 'dep...   
1  ['preheat oven to 425 degrees f', 'press dough...   
2  ['brown ground beef in large pot', 'add choppe...   
3  ['place potatoes in a large pot of lightly sal...   
4  ['mix all ingredients& boil for 2 1 / 2 hours ...   

                                         description  \
0  autumn is my favorite time of year to cook! th...   
1  this recipe calls for the crust to be prebaked...   
2  this modified version of 'mom's' chili was a h...   
3  this is a super easy, great tasting, make ahea...   
4  my dh's amish mother raised him on this recipe...   

                                         ingredients  n_ingredients  
0  ['winter squash', 'mexican seasoning', 'mixed ...              7  
1  ['prepared pizza crust', 'sausage patty', 'egg...              6  
2  ['ground beef', 'yellow onions', 'diced tomato...             13  
3  ['spreadable cheese with garlic and herbs', 'n...             11  
4  ['tomato juice', 'apple cider vinegar', 'sugar...              8

We want to see how an entry of "tags" column of this file looks like.

We will write a function that turns a string to a list. We will use this function to turn some string entries of the columns RAW_recipes into lists. 

In [223]:
import ast

def convert_to_list(tags_str):
    try:
        return ast.literal_eval(tags_str)
    except (ValueError, SyntaxError):
        return []

In [224]:
RAW_recipes['tags']=RAW_recipes['tags'].apply(convert_to_list)

In [225]:
RAW_recipes['tags'][100]

['15-minutes-or-less',
 'time-to-make',
 'course',
 'main-ingredient',
 'cuisine',
 'preparation',
 'occasion',
 'for-1-or-2',
 'low-protein',
 'healthy',
 'lunch',
 'salads',
 'snacks',
 'vegetables',
 'asian',
 'indian',
 'easy',
 'beginner-cook',
 'low-fat',
 'summer',
 'vegetarian',
 'dietary',
 'spicy',
 'low-sodium',
 'low-cholesterol',
 'seasonal',
 'low-saturated-fat',
 'low-calorie',
 'low-carb',
 'inexpensive',
 'healthy-2',
 'low-in-something',
 'tomatoes',
 'taste-mood',
 'savory',
 'number-of-servings',
 'presentation',
 'served-cold']

We will write a function that will help us to read the "tags" column of RAW_recipes and look for certain cuisine related keywords. The following is keyword for the 100 best cuisines in the world given by Taste Atlas.

In [226]:
world_cuisines = ['italian',
    'japanese',
    'greek',
    'portuguese',
    'chinese',
    'indonesian',
    'mexican',
    'french',
    'spanish',
    'peruvian',
    'indian',
    'brazilian',
    'polish',
    'argentinian',
    'turkish',
    'states',
    'thai',
    'korean',
    'croatian',
    'serbian',
    'hungarian',
    'vietnamese',
    'iranian',
    'chilean',
    'lebanese',
    'georgian',
    'bosnian',
    'colombian',
    'romanian',
    'bulgarian',
    'malaysian',
    'german',
    'filipino',
    'belgian',
    'czech',
    'austrian',
    'swiss',
    'lithuanian',
    'english',
    'algerian',
    'russian',
    'slovak',
    'canadian',
    'swedish',
    'dutch',
    'moroccan',
    'scottish',
    'ecuadorian',
    'danish',
    'australian',
    'egyptian',
    'south-african',
    'ukrainian',
    'syrian',
    'irish',
    'singaporean',
    'pakistani',
    'puerto',
    'norwegian',
    'bolivian',
    'macedonian',
    'israeli',
    'palestinian',
    'slovenian',
    'finnish',
    'tunisian',
    'haitian',
    'jamaican',
    'armenian',
    'venezuelan',
    'belarusian',
    'moldovan',
    'lankan',
    'jordanian',
    'cuban',
    'uzbekistani',
    'azerbaijani',
    'taiwanese',
    'uruguayan',
    'montenegrin',
    'ethiopian',
    'iraqi',
    'qatari',
    'trinidadian',
    'libyan',
    'lao',
    'barbadian',
    'cypriot',
    'bengali',
    'kazakhstani',
    'albanian',
    'kyrgyzstani',
    'burmese',
    'zealand',
    'saudi',
    'irish',
    'bahamian',
    'dominican',
    'welsh',
    'ghanaian'
]

We will write a function that searches entries of "tags" column of the RAW_recipes for the keywords above. If there is a string which has the keyowrds above we return them.

In [227]:
def extract_cuisine_tags(tags):
    cuisine_related_tags = []
    for tag in tags:
        for cuisine in world_cuisines:
            if cuisine in tag:
                cuisine_related_tags.append(tag)
    return cuisine_related_tags

# Create a new column in the RAW_recipes with the extracted cuisine-related tags
RAW_recipes['Cuisine_Tags'] = RAW_recipes['tags'].apply(extract_cuisine_tags)

In [228]:
RAW_recipes_cuisine_cleaned.head()

name      id  minutes  \
0  arriba   baked winter squash mexican style  137739       55   
1            a bit different  breakfast pizza   31490       30   
2          amish  tomato ketchup  for canning   44061      190   
3                       aww  marinated olives   25274       15   
4                              chile rellenos   43026       45   

   contributor_id   submitted  \
0           47892  2005-09-16   
1           26278  2002-06-17   
2           41706  2002-10-25   
3           21730  2002-04-14   
4           52268  2002-10-14   

                                                tags  \
0  [60-minutes-or-less, time-to-make, course, mai...   
1  [30-minutes-or-less, time-to-make, course, mai...   
2  [weeknight, time-to-make, course, main-ingredi...   
3  [15-minutes-or-less, time-to-make, course, mai...   
4  [60-minutes-or-less, time-to-make, course, mai...   

                                   nutrition  n_steps  \
0      [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]       11   
1  [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]        9   
2  [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]        5   
3   [380.7, 53.0, 7.0, 24.0, 6.0, 24.0, 6.0]        4   
4   [94.0, 10.0, 0.0, 11.0, 11.0, 21.0, 0.0]        9   

                                               steps  \
0  ['make a choice and proceed with recipe', 'dep...   
1  ['preheat oven to 425 degrees f', 'press dough...   
2  ['mix all ingredients& boil for 2 1 / 2 hours ...   
3  ['toast the fennel seeds and lightly crush the...   
4  ['drain green chiles', 'sprinkle cornstarch on...   

                                         description  \
0  autumn is my favorite time of year to cook! th...   
1  this recipe calls for the crust to be prebaked...   
2  my dh's amish mother raised him on this recipe...   
3  my italian mil was thoroughly impressed by my ...   
4  a favorite from a local restaurant no longer i...   

                                         ingredients  n_ingredients  \
0  ['winter squash', 'mexican seasoning', 'mixed ...              7   
1  ['prepared pizza crust', 'sausage patty', 'egg...              6   
2  ['tomato juice', 'apple cider vinegar', 'sugar...              8   
3  ['fennel seeds', 'green olives', 'ripe olives'...              9   
4  ['egg roll wrap', 'whole green chilies', 'chee...              5   

                   Cuisine_Tags  
0                     [mexican]  
1  [northeastern-united-states]  
2  [northeastern-united-states]  
3                    [canadian]  
4  [southwestern-united-states]

Now we will try to understand how many recipies are there with a Cuisine Tag we created.

In [230]:
# Count the occurrences of each category in the Cuisine_Tags column
cuisine_category_counts = RAW_recipes['Cuisine_Tags'].value_counts()

# Print the counts of all categories for reference
print(cuisine_category_counts)

Cuisine_Tags
[]                                                                            176104
[italian]                                                                       6911
[southern-united-states]                                                        6277
[mexican]                                                                       5642
[canadian]                                                                      3923
                                                                               ...  
[canadian, french, southern-united-states, russian]                                1
[brazilian, mexican, peruvian, venezuelan]                                         1
[southern-united-states, irish, irish, polish, northeastern-united-states]         1
[southwestern-united-states, moroccan]                                             1
[greek, moroccan]                                                                  1
Name: count, Length: 667, dtype: int64


For the training purposes, we will clean the ones without a Cuisine Tag and the ones with more than a Cuisine Tag.

In [231]:
# Get the indices of rows that do not have exactly one cuisine-related tag
indices_no_cuisine_tags = RAW_recipes[RAW_recipes['Cuisine_Tags'].apply(lambda x: len(x) != 1)].index

# Drop them
RAW_recipes_cuisine_cleaned = RAW_recipes.drop(indices_no_cuisine_tags).reset_index(drop=True)

In [232]:
# Count the occurrences of each category in the Cuisine_Tags
cusine_category_counts = RAW_recipes_cuisine_cleaned['Cuisine_Tags'].value_counts()

# Print the counts of all categories for reference
print(cusine_category_counts)

Cuisine_Tags
[italian]                       6911
[southern-united-states]        6277
[mexican]                       5642
[canadian]                      3923
[indian]                        2564
[australian]                    2355
[southwestern-united-states]    2194
[greek]                         2165
[chinese]                       1841
[french]                        1813
[northeastern-united-states]    1724
[german]                        1186
[english]                       1173
[thai]                          1116
[spanish]                        871
[moroccan]                       839
[japanese]                       768
[new-zealand]                    405
[scottish]                       381
[swedish]                        340
[russian]                        322
[vietnamese]                     314
[south-african]                  313
[swiss]                          302
[polish]                         299
[portuguese]                     292
[hungarian]              

There are 6 recipes with a tag "chinese-new-year". We will turn them into just "chinese".

In [233]:
# Define a function to replace 'chinese-new-year' with 'chinese'
def replace_chinese_new_year(tags):
    return ['chinese' if tag == 'chinese-new-year' else tag for tag in tags]

# Apply the function to the 'Cuisine_Tags' column
RAW_recipes_cuisine_cleaned['Cuisine_Tags'] = RAW_recipes_cuisine_cleaned['Cuisine_Tags'].apply(replace_chinese_new_year)

Now we will check if everything went as we wanted.

In [234]:
# Count the occurrences of each category in the Time_Category column
cusine_category_counts = RAW_recipes_cuisine_cleaned['Cuisine_Tags'].value_counts()

# Print the counts of all categories for reference
print(cusine_category_counts)

Cuisine_Tags
[italian]                       6911
[southern-united-states]        6277
[mexican]                       5642
[canadian]                      3923
[indian]                        2564
[australian]                    2355
[southwestern-united-states]    2194
[greek]                         2165
[chinese]                       1847
[french]                        1813
[northeastern-united-states]    1724
[german]                        1186
[english]                       1173
[thai]                          1116
[spanish]                        871
[moroccan]                       839
[japanese]                       768
[new-zealand]                    405
[scottish]                       381
[swedish]                        340
[russian]                        322
[vietnamese]                     314
[south-african]                  313
[swiss]                          302
[polish]                         299
[portuguese]                     292
[hungarian]              

In [99]:
len(RAW_recipes_cuisine_cleaned)

50399

In [100]:
len(cusine_category_counts)

58

So, we have categorized 50399 recipes into 58 cuisine categories. We can use this data set as a training set for a model that predicts the cuisine for a given list of ingredients or a recipe.

In [101]:
# Write the DataFrame to a CSV file
RAW_recipes_cuisine_cleaned.to_csv('RAW_recipes_cuisine_cleaned.csv', index=False)

# Verify the file creation by printing a message
print("DataFrame has been written to RAW_recipes_cuisine_cleaned.csv")

DataFrame has been written to RAW_recipes_cuisine_cleaned.csv


In [203]:
# Count the occurrences of each category in the Cuisine_Tags column
cuisine_category_counts = RAW_recipes_cuisine_cleaned['Cuisine_Tags'].value_counts()

# Identify cuisines with at least 101 occurrences
cuisines_to_keep = cuisine_category_counts[cuisine_category_counts >= 400].index

In [204]:
cuisines_to_keep_list = cuisines_to_keep.tolist()

In [205]:
cuisines_to_keep_list

[['italian'],
 ['southern-united-states'],
 ['mexican'],
 ['canadian'],
 ['indian'],
 ['australian'],
 ['southwestern-united-states'],
 ['greek'],
 ['chinese'],
 ['french'],
 ['northeastern-united-states'],
 ['german'],
 ['english'],
 ['thai'],
 ['spanish'],
 ['moroccan'],
 ['japanese'],
 ['new-zealand']]

In [206]:
filtered_recipes = RAW_recipes_cuisine_cleaned[RAW_recipes_cuisine_cleaned['Cuisine_Tags'].apply(lambda x: x in cuisines_to_keep_list)]

In [207]:
filtered_recipes.head()

name      id  minutes  \
0  arriba   baked winter squash mexican style  137739       55   
1            a bit different  breakfast pizza   31490       30   
2          amish  tomato ketchup  for canning   44061      190   
3                       aww  marinated olives   25274       15   
4                              chile rellenos   43026       45   

   contributor_id   submitted  \
0           47892  2005-09-16   
1           26278  2002-06-17   
2           41706  2002-10-25   
3           21730  2002-04-14   
4           52268  2002-10-14   

                                                tags  \
0  [60-minutes-or-less, time-to-make, course, mai...   
1  [30-minutes-or-less, time-to-make, course, mai...   
2  [weeknight, time-to-make, course, main-ingredi...   
3  [15-minutes-or-less, time-to-make, course, mai...   
4  [60-minutes-or-less, time-to-make, course, mai...   

                                   nutrition  n_steps  \
0      [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]       11   
1  [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]        9   
2  [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]        5   
3   [380.7, 53.0, 7.0, 24.0, 6.0, 24.0, 6.0]        4   
4   [94.0, 10.0, 0.0, 11.0, 11.0, 21.0, 0.0]        9   

                                               steps  \
0  ['make a choice and proceed with recipe', 'dep...   
1  ['preheat oven to 425 degrees f', 'press dough...   
2  ['mix all ingredients& boil for 2 1 / 2 hours ...   
3  ['toast the fennel seeds and lightly crush the...   
4  ['drain green chiles', 'sprinkle cornstarch on...   

                                         description  \
0  autumn is my favorite time of year to cook! th...   
1  this recipe calls for the crust to be prebaked...   
2  my dh's amish mother raised him on this recipe...   
3  my italian mil was thoroughly impressed by my ...   
4  a favorite from a local restaurant no longer i...   

                                         ingredients  n_ingredients  \
0  ['winter squash', 'mexican seasoning', 'mixed ...              7   
1  ['prepared pizza crust', 'sausage patty', 'egg...              6   
2  ['tomato juice', 'apple cider vinegar', 'sugar...              8   
3  ['fennel seeds', 'green olives', 'ripe olives'...              9   
4  ['egg roll wrap', 'whole green chilies', 'chee...              5   

                   Cuisine_Tags  
0                     [mexican]  
1  [northeastern-united-states]  
2  [northeastern-united-states]  
3                    [canadian]  
4  [southwestern-united-states]

In [208]:
type(filtered_recipes)

pandas.core.frame.DataFrame

In [214]:
# Write the DataFrame to a CSV file
filtered_recipes.to_csv('RAW_recipes_cuisine_cleaned_2.csv', index=False)

# Verify the file creation by printing a message
print("DataFrame has been written to RAW_recipes_cuisine_cleaned_2.csv")

DataFrame has been written to RAW_recipes_cuisine_cleaned_2.csv


In [212]:
# Count the occurrences of each category in the Time_Category column
cusine_category_counts = filtered_recipes['Cuisine_Tags'].value_counts()

# Print the counts of all categories for reference
print(cusine_category_counts)

Cuisine_Tags
[united_states]    10195
[italian]           6911
[mexican]           5642
[canadian]          3923
[indian]            2564
[australian]        2355
[greek]             2165
[chinese]           1847
[french]            1813
[german]            1186
[english]           1173
[thai]              1116
[spanish]            871
[moroccan]           839
[japanese]           768
[new-zealand]        405
Name: count, dtype: int64


In [213]:
len(cusine_category_counts)

16

In [211]:
# Define a function to replace 'chinese-new-year' with 'chinese'
def united_states_cuisine(tags):
    return ['united_states' if tag in ['southern-united-states','southwestern-united-states','northeastern-united-states'] else tag for tag in tags]

# Apply the function to the 'Cuisine_Tags' column
filtered_recipes['Cuisine_Tags'] = filtered_recipes['Cuisine_Tags'].apply(united_states_cuisine)

C:\Users\Deniz\AppData\Local\Temp\ipykernel_11588\2766634347.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_recipes['Cuisine_Tags'] = filtered_recipes['Cuisine_Tags'].apply(united_states_cuisine)


In [137]:
filtered_recipes['ingredients']=filtered_recipes['ingredients'].apply(convert_to_list)

C:\Users\Deniz\AppData\Local\Temp\ipykernel_11588\2464868034.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_recipes['ingredients']=filtered_recipes['ingredients'].apply(convert_to_list)


In [141]:
A=filtered_recipes['ingredients']

In [145]:
A

0        [winter squash, mexican seasoning, mixed spice...
1        [prepared pizza crust, sausage patty, eggs, mi...
2        [tomato juice, apple cider vinegar, sugar, sal...
3        [fennel seeds, green olives, ripe olives, garl...
4        [egg roll wrap, whole green chilies, cheese, c...
                               ...                        
50394               [butter, onions, flour, salt, vinegar]
50395    [onions, red onions, fresh garlic, olive oil, ...
50396    [flour, baking powder, kosher salt, unsalted b...
50397    [mayonnaise, prepared horseradish, worcestersh...
50398    [celery, onion, green sweet pepper, garlic clo...
Name: ingredients, Length: 47903, dtype: object

In [146]:
all_ingredients=set()
# Iterate over each dictionary in the list and update the set with ingredients
for ingredients_list in A:
    all_ingredients.update(ingredients_list)

# Convert the set back to a list
all_ingredients_list = list(all_ingredients)

In [150]:
all_ingredients

{'sweet whole kernel corn',
 'roasted peanuts',
 'kraft zesty italian dressing',
 'beef burgers',
 'strawberry glaze',
 'unsweetened cocoa',
 'canned jalapeno pepper',
 'watercress',
 'paper towels',
 'ham spread',
 'dandelion flowers',
 'beet juice',
 "goat's milk",
 'yellowtail fish fillets',
 'condensed chicken gumbo soup',
 'ajwain',
 'fresh scallops',
 'hummus',
 'plain soymilk',
 'white corn tortilla',
 'black sesame seeds',
 "red bird's eye chili",
 "campbell's cream of chicken soup",
 'boneless beef chuck roast',
 'diced onion',
 'refined flour',
 'wheat starch',
 'instant polenta',
 'cauliflower florets',
 'nashi',
 'orange slice',
 'pears in heavy syrup',
 'red wine vinegar',
 'fish bouillon cube',
 'condensed beef broth',
 'garbanzo beans',
 'dry lentils',
 'strawberry puree',
 'lean flank steak',
 'corn chowder',
 'low-fat vanilla yogurt',
 'trout fillet',
 'salad dressing',
 'butterfly pork chops',
 'bean curd',
 'black licorice strings',
 'sesame oil',
 'diced tomatoes',


In [152]:
# Convert the set back to a list
all_ingredients_list = list(all_ingredients)

# Sort the list alphabetically
all_ingredients_list_sorted = sorted(all_ingredients_list)

# Create a matrix where each row is a binary vector for each entry
binary_matrix = []
for ingredients_list in A:
    vector = [1 if ingredient in ingredients_list else 0 for ingredient in all_ingredients_list_sorted]
    binary_matrix.append(vector)

In [168]:
import numpy as np
use_of_elements = np.sum(binary_matrix, axis=0)

In [177]:
indices_of_freq_ingr=np.argsort(use_of_elements)[-50:]

In [178]:
indices_of_freq_ingr

array([1359, 7137, 2352, 1355, 2721, 4392, 6790, 8896, 2159, 3324, 1387,
       1824, 1446, 3367, 6230, 3413, 1531, 6218, 1393, 3767, 8088, 1643,
       3015, 4124,  386, 8047, 1022, 3755, 6019, 5017, 6070, 2935,  118,
        686, 9061, 6224, 7542, 8735, 5670, 3228, 6359, 3738, 2955, 3756,
       9178, 8311, 6040, 1069, 6049, 7536], dtype=int64)

In [179]:
frequent_ingredients=[]
for i in range(0,len(indices_of_freq_ingr)-1):frequent_ingredients.append(all_ingredients_list_sorted[indices_of_freq_ingr[i]])

In [180]:
frequent_ingredients

['carrots',
 'red onion',
 'cumin',
 'carrot',
 'dried oregano',
 'honey',
 'potatoes',
 'unsalted butter',
 'cornstarch',
 'fresh cilantro',
 'cayenne pepper',
 'cinnamon',
 'cheddar cheese',
 'fresh ground black pepper',
 'parsley',
 'fresh parsley',
 'chicken broth',
 'paprika',
 'celery',
 'garlic powder',
 'soy sauce',
 'chili powder',
 'extra virgin olive oil',
 'ground cumin',
 'baking powder',
 'sour cream',
 'brown sugar',
 'garlic clove',
 'oil',
 'lemon juice',
 'onions',
 'egg',
 'all-purpose flour',
 'black pepper',
 'vegetable oil',
 'parmesan cheese',
 'salt and pepper',
 'tomatoes',
 'milk',
 'flour',
 'pepper',
 'garlic',
 'eggs',
 'garlic cloves',
 'water',
 'sugar',
 'olive oil',
 'butter',
 'onion']

In [176]:
use_of_elements[6049]

10347

In [182]:
import inflect
import nltk
from nltk.tokenize import word_tokenize
from nltk.tag import pos_tag

# Download NLTK data files (only needed once)
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')

# Create an inflect engine instance
p = inflect.engine()

def ensure_singular(phrase):
    # Tokenize the phrase into words
    words = word_tokenize(phrase)
    
    # Perform part-of-speech tagging
    pos_tags = pos_tag(words)
    
    # If the phrase has more than one word, return it as is
    if len(words) > 1:
        return phrase
    
    # Check if the single word is already singular
    word = words[0]
    if not p.singular_noun(word):  # If the word is not plural
        return phrase  # It's already singular, return as is
    
    # Otherwise, convert it to singular
    return p.singular_noun(word)

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Deniz\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\Deniz\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping taggers\averaged_perceptron_tagger.zip.


['carrot',
 'red onion',
 'cumin',
 'carrot',
 'dried oregano',
 'honey',
 'potato',
 'unsalted butter',
 'cornstarch',
 'fresh cilantro',
 'cayenne pepper',
 'cinnamon',
 'cheddar cheese',
 'fresh ground black pepper',
 'parsley',
 'fresh parsley',
 'chicken broth',
 'paprika',
 'celery',
 'garlic powder',
 'soy sauce',
 'chili powder',
 'extra virgin olive oil',
 'ground cumin',
 'baking powder',
 'sour cream',
 'brown sugar',
 'garlic clove',
 'oil',
 'lemon juice',
 'onion',
 'egg',
 'all-purpose flour',
 'black pepper',
 'vegetable oil',
 'parmesan cheese',
 'salt and pepper',
 'tomato',
 'milk',
 'flour',
 'pepper',
 'garlic',
 'egg',
 'garlic cloves',
 'water',
 'sugar',
 'olive oil',
 'butter',
 'onion']